source/material_model/rheology/friction_options.cc

```diff
@@ -0,0 +1,849 @@
+/*
+  Copyright (C) 2019 - 2020 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <aspect/material_model/rheology/friction_options.h>
+
+#include <deal.II/base/signaling_nan.h>
+#include <deal.II/base/parameter_handler.h>
+#include <aspect/utilities.h>
+#include <aspect/geometry_model/interface.h>
+
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/base/quadrature_lib.h>
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    namespace Rheology
+    {
+      template <int dim>
+      double
+      FrictionOptions<dim>::
+      compute_dependent_friction_angle(const double current_edot_ii,
+                                       const unsigned int j,  // j is from a for-loop over volume fractions
+                                       const std::vector<double> &composition,
```

Comment:
So is composition the volume fractions?

Reply:
I hand over in.composition[i]. As I use it together with the volume_fractions index j, I guess in.composition also includes the background field and includes all original compositions not only volume_fractions. This is in analogy to the other parameters within calculate_isostrain_viscosities. But now that you asked this question, I am wondering in general why we loop over volume_fractions.size() and not over the non-masked compositions only, which should be faster. Is there a reason? It is done everywhere in visco-plastic as far as I remember. The masked fields like stress_xx or theta should just return a 0 anyway, as they are not part of the volume? So why do we loop over all fields?

Reply:
`in.composition[i]` does not have the background field, but does indeed include all the 'raw' compositional field values. The index `j` is from the volume fractions, so it should not be directly used on `in.composition[i]`. You need `in.composition[i]` to get at theta, but as long as you use `theta_composition_index` I think it's fine. But do add a comment that `composition` is not volume_fractions but the compositional fields. 

I agree that instead of having volume_fractions = n_fields + 1, it could be n_fields + 1 - masked fields. This would require restructuring of the input logic (not specifying material properties for the masked fields) and the compute_volume_fractions function. The masked fields came way later than these structures, so I think it's just grown this way. It would break compatibility, but something to discuss!